### PR TITLE
feat: add a configuration option for maximum message size in zeebe client java

### LINKED ISSUE #12122: Configure the client's inbound max_message_size
**Is your feature request related to a problem? Please describe.**
Sometimes, I may want to specify a larger than the default (4 MB) inbound `MAX_MESSAGE_SIZE` for the client. For example, when I create a process instance with a result where the resulting variables are larger than the default 4 MB.

**Describe the solution you'd like**
Add a configuration option to the client to specify an inbound max message size applied to the gRPC responses. The default should be the current hardcoded 4 MB.

**Describe alternatives you've considered**
- specify it for each command: this is hard to implement at this time as the channel is set up on client construction
- don't allow specifying it: the current behavior throws errors in the client when the gateway responds with larger than 4MB messages (which is possible already).

**Additional context**
- https://github.com/camunda/zeebe/pull/11902#issuecomment-1480460031
- https://github.com/camunda/zeebe/issues/12104

### PR DESCRIPTION
**Pull request by gakkiyomi**
## Description

Allows configuring the Java client's inbound max message size, which specifies the maximum size of incoming responses the client can receive. Specifically, it sets the `maxInboundMessageSize` of the gRPC channel. 

The default is 4194304 = 4MB.

It can be configured:
- as an environment variables `zeebe.client.maxMessageSize` (as data size string)
- using a configuration property: `zeebe.client.maxMessageSize` (as data size string)
- or using the Java API: `ZeebeClientBuilder.maxMessageSize(int maxMessageSize)`

A data size string is a number optionally followed by a data size unit `kb|KB|mb|MB`. For example: `4MB`, `4194304`, `4000 kb`, etc.

## Related issues

closes #12122 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [x] The change has been verified by a QA run
* [x] The impact of the changes is verified by a benchmark

Documentation:
* [x] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by korthout — CHANGES_REQUESTED** [review]
Hi @gakkiyomi. First off, my sincere apologies for awaiting a review for so long 🙇 

But of course, thank you for contributing to Zeebe 👏 

❌ I like the code you wrote for this, but I cannot accept it without considering the issue it tries to resolve.

❓ Can you explain why you think this change is needed? What is your use case, and why isn't the [max message size](https://github.com/camunda/zeebe/blob/main/dist/src/main/config/broker.standalone.yaml.template#L135-L137) defined by Zeebe's Broker enough? Why should the client allow defining an inbound max message size that differs from the Broker's max message size?

🏗️ We appreciate your contribution. [Please always consider discussing a feature request in an issue before opening a pull request in the future](https://github.com/camunda/zeebe/blob/main/CONTRIBUTING.md#starting-on-an-issue), to avoid that we have to reject it because it's unclear whether the change can be accepted.

**Comment by gakkiyomi** [comment]
> 

Thank you for your reply. I encountered a client-side restriction when transmitting messages larger than 4m. At the same time, I noticed that the broker also supports setting maxMessageSize in the NetworkCfg class and can even be configured through the environment variable zeebe broker.network.maxMessageSize. I want to implement this feature for clients as well, which is why this PR was created.

**Comment by korthout** [comment]
@gakkiyomi Thanks for clarifying. That makes sense to me.

By coincidence, today, [I showed](https://github.com/camunda/zeebe/issues/12104#issuecomment-1480916861) that I could create a process instance with variables exceeding the default 4 MB. 

But in that example, the client does not receive a message with such a large size because the variables that the process instance is created with are not responded to the client. The client's inbound MAX_MESSAGE_SIZE is not challenged.

Switching that same example to use `.withResult()` and checking the variables in the response fails with an error, because now the client's inbound `MAX_MESSAGE_SIZE` is challenged.

```java
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Zeebe Community License 1.1. You may not use this file
 * except in compliance with the Zeebe Community License 1.1.
 */
package io.camunda.zeebe.it.client.command;

import static org.assertj.core.api.Assertions.assertThat;

import io.camunda.zeebe.broker.test.EmbeddedBrokerRule;
import io.camunda.zeebe.client.api.response.ProcessInstanceResult;
import io.camunda.zeebe.it.util.GrpcClientRule;
import io.camunda.zeebe.model.bpmn.Bpmn;
import io.camunda.zeebe.test.util.BrokerClassRuleHelper;
import io.camunda.zeebe.util.ByteValue;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import org.junit.ClassRule;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.RuleChain;
import org.springframework.util.unit.DataSize;

public final class CreateProcessInstanceTest2 {

  private static final EmbeddedBrokerRule BROKER_RULE =
      new EmbeddedBrokerRule(
          cfg -> {
            // set MAX_MESSAGE_SIZE to twice the maximum variable size + slight more
            cfg.getNetwork().setMaxMessageSize(DataSize.ofMegabytes(60));
            cfg.getGateway().getNetwork().setMaxMessageSize(DataSize.ofMegabytes(60));
          });
  private static final GrpcClientRule CLIENT_RULE = new GrpcClientRule(BROKER_RULE);

  @ClassRule
  public static RuleChain ruleChain = RuleChain.outerRule(BROKER_RULE).around(CLIENT_RULE);

  @Rule public final BrokerClassRuleHelper helper = new BrokerClassRuleHelper();

  @Test
  public void shouldCreateInstanceWithLargeVariables() {
    // given
    CLIENT_RULE.deployProcess(
        Bpmn.createExecutableProcess("PROCESS").startEvent().endEvent().done());

    // when
    final ProcessInstanceResult processInstance =
        CLIENT_RULE
            .getClient()
            .newCreateInstanceCommand()
            .bpmnProcessId("PROCESS")
            .latestVersion()
            .variables(Map.of("variable", "x".repeat((int) (ByteValue.ofMegabytes(27)))))
            .withResult()
            .send()
            .join();

    // then
    assertThat(processInstance.getVariables().getBytes(StandardCharsets.UTF_8))
        .hasSizeGreaterThan((int) ByteValue.ofMegabytes(27));
  }
}
```

Please add such a test case to your pull request. (perhaps we lower the variable sizes a bit as 27MB seems excessive).

I've opened [a feature request](https://github.com/camunda/zeebe/issues/12122) for this, which we can refer to in the pull request description.

**Comment by gakkiyomi** [comment]
> @gakkiyomi Thanks for clarifying. That makes sense to me.
> 
> By coincidence, today, [I showed](https://github.com/camunda/zeebe/issues/12104#issuecomment-1480916861) that I could create a process instance with variables exceeding the default 4 MB.
> 
> But in that example, the client does not receive a message with such a large size because the variables that the process instance is created with are not responded to the client. The client's inbound MAX_MESSAGE_SIZE is not challenged.
> 
> Switching that same example to use `.withResult()` and checking the variables in the response fails with an error, because now the client's inbound `MAX_MESSAGE_SIZE` is challenged.
> 
> ```java
> /*
>  * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
>  * one or more contributor license agreements. See the NOTICE file distributed
>  * with this work for additional information regarding copyright ownership.
>  * Licensed under the Zeebe Community License 1.1. You may not use this file
>  * except in compliance with the Zeebe Community License 1.1.
>  */
> package io.camunda.zeebe.it.client.command;
> 
> import static org.assertj.core.api.Assertions.assertThat;
> 
> import io.camunda.zeebe.broker.test.EmbeddedBrokerRule;
> import io.camunda.zeebe.client.api.response.ProcessInstanceResult;
> import io.camunda.zeebe.it.util.GrpcClientRule;
> import io.camunda.zeebe.model.bpmn.Bpmn;
> import io.camunda.zeebe.test.util.BrokerClassRuleHelper;
> import io.camunda.zeebe.util.ByteValue;
> import java.nio.charset.StandardCharsets;
> import java.util.Map;
> import org.junit.ClassRule;
> import org.junit.Rule;
> import org.junit.Test;
> import org.junit.rules.RuleChain;
> import org.springframework.util.unit.DataSize;
> 
> public final class CreateProcessInstanceWithLargeResultTest {
> 
>   private static final EmbeddedBrokerRule BROKER_RULE =
>       new EmbeddedBrokerRule(
>           cfg -> {
>             // set MAX_MESSAGE_SIZE to twice the maximum variable size + slight more
>             cfg.getNetwork().setMaxMessageSize(DataSize.ofMegabytes(60));
>             cfg.getGateway().getNetwork().setMaxMessageSize(DataSize.ofMegabytes(60));
>           });
>   private static final GrpcClientRule CLIENT_RULE = new GrpcClientRule(BROKER_RULE);
> 
>   @ClassRule
>   public static RuleChain ruleChain = RuleChain.outerRule(BROKER_RULE).around(CLIENT_RULE);
> 
>   @Rule public final BrokerClassRuleHelper helper = new BrokerClassRuleHelper();
> 
>   @Test
>   public void shouldCreateInstanceWithLargeResult() {
>     // given
>     CLIENT_RULE.deployProcess(
>         Bpmn.createExecutableProcess("PROCESS").startEvent().endEvent().done());
> 
>     // when
>     final ProcessInstanceResult processInstance =
>         CLIENT_RULE
>             .getClient()
>             .newCreateInstanceCommand()
>             .bpmnProcessId("PROCESS")
>             .latestVersion()
>             .variables(Map.of("variable", "x".repeat((int) (ByteValue.ofMegabytes(27)))))
>             .withResult()
>             .send()
>             .join();
> 
>     // then
>     assertThat(processInstance.getVariables().getBytes(StandardCharsets.UTF_8))
>         .hasSizeGreaterThan((int) ByteValue.ofMegabytes(27));
>   }
> }
> ```
> 
> Please add such a test case to your pull request. (perhaps we lower the variable sizes a bit as 27MB seems excessive).
> 
> I've opened [a feature request](https://github.com/camunda/zeebe/issues/12122) for this, which we can refer to in the pull request description.

@korthout Thank you for your suggestion. I have added a test case for this feature, located at `qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CreateProcessInstanceWithLargeResultTest.java`

File: clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClientBuilder.java
Comment (korthout) [inline clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClientBuilder.java:121]: ❌ We'll need to add documentation to clarify to users what happens when they set it and what the default is. It should also be clear what unit is expected.

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java
Comment (korthout) [inline clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java:24]: 🔧 We can simplify this pattern to only fit:
- numbers without datasize unit (considered bytes)
- numbers with datasize unit (either `kb` or `mb`)

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java
Comment (korthout) [inline clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java:36]: We have some [guidelines for errors](https://github.com/camunda/zeebe/wiki/Error-Guidelines) to ensure they help users figure out how to resolve the problem.

❌ Please apply the guidelines

For example, you could use something like `Expected '${text}' to be a data size, but does not match pattern '${pattern}'`

File: clients/java/src/test/java/io/camunda/zeebe/client/ZeebeClientTest.java
Comment (korthout) [inline clients/java/src/test/java/io/camunda/zeebe/client/ZeebeClientTest.java:232]: ❌ There's one test case missing: `shouldOverrideMaxMessageSizeWithEnvVar`

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks @gakkiyomi, for your changes 🏗️ 

👍 The main parts are in place! I like it!

❌ I've had a more thorough look over the changes, and there are some things that must be improved before we can merge this.

Please have a look at my comments 🙇

File: clients/java/src/test/java/io/camunda/zeebe/client/ZeebeClientTest.java
Comment (gakkiyomi) [reply clients/java/src/test/java/io/camunda/zeebe/client/ZeebeClientTest.java:245]: > ❌ There's one test case missing: `shouldOverrideMaxMessageSizeWithEnvVar`

fixed

File: clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClientBuilder.java
Comment (gakkiyomi) [reply clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClientBuilder.java:125]: fixed

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java
Comment (gakkiyomi) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java]: fixed

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java
Comment (gakkiyomi) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java]: fixed

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java
Comment (gakkiyomi) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java]: fixed

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java
Comment (gakkiyomi) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java]: fixed

**Comment by gakkiyomi** [comment]
@korthout Thank you for taking the time to review my PR again. I have fixed the issues you pointed out.

File: clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClientBuilder.java
Comment (korthout) [inline clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClientBuilder.java:125]: 🔧 This documentation is quite technical. We could consider adding what a user gains by setting it. For example, something like 

> A custom maxMessageSize allows the client to receive larger or smaller responses from Zeebe. Technically, it specifies the maxInboundMessageSize of the gRPC channel. The default is 4194304 = 4MB.

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java
Comment (korthout) [inline clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java:49]: 🔧 Please apply the error guidelines to all thrown exceptions in this class so that users know how to resolve the problem when confronted with that error message.

File: clients/java/src/test/java/io/camunda/zeebe/client/ZeebeClientTest.java
Comment (korthout) [inline clients/java/src/test/java/io/camunda/zeebe/client/ZeebeClientTest.java:135]: 🔧 We should aim to keep the test classes consistently structured. Please move this test case next to the other maxMessageSize test cases.

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CreateProcessInstanceWithLargeResultTest.java
Comment (korthout) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CreateProcessInstanceWithLargeResultTest.java:57]: 🔧 If we expect a response to this command using a variable of 10MB, then we can lower the used maxMessageSize to 21 MB or so. Currently, the comment with the configuration doesn't help make things easier to understand anymore:

```java
// set MAX_MESSAGE_SIZE to twice the maximum variable size + slight more
cfg.getNetwork().setMaxMessageSize(DataSize.ofMegabytes(21));
cfg.getGateway().getNetwork().setMaxMessageSize(DataSize.ofMegabytes(21));
```

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java
Comment (korthout) [inline clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java:25]: 👍 Nice! This pattern is much easier to understand

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks for applying the requested changes so fast 💯 

👍 It looks great, we're almost there

❌ Please make sure to apply [the checkstyle hints](https://github.com/camunda/zeebe/actions/runs/4571263144/jobs/8076381022?pr=11902)

❌ Please remove the merge commit 0534e43bbfff6d9266a53442fd1b0bdeaeb957c8, and instead rebase your changes on top of `main` (this keeps our git history easier to navigate)

❌ Please add a description of the feature to the pull request description, and link issue #12122 as closes (so our release process can label that issue correctly when this feature is released)

PS: comments marked as 🔧 are suggestions. I would also accept the changes as they are. See [review emoji code](https://github.com/camunda/zeebe/blob/main/CONTRIBUTING.md#review-emoji-code)

File: clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClientBuilder.java
Comment (gakkiyomi) [reply clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClientBuilder.java:125]: fixed

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java
Comment (gakkiyomi) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/util/DataSizeUtil.java:49]: fixed

File: clients/java/src/test/java/io/camunda/zeebe/client/ZeebeClientTest.java
Comment (gakkiyomi) [reply clients/java/src/test/java/io/camunda/zeebe/client/ZeebeClientTest.java:135]: fixed

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CreateProcessInstanceWithLargeResultTest.java
Comment (gakkiyomi) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CreateProcessInstanceWithLargeResultTest.java:57]: fixed

**Comment by gakkiyomi** [comment]
@korthout I have revised the suggestion, please review it.

**Comment by korthout** [comment]
@gakkiyomi Please resolve the errors in https://github.com/camunda/zeebe/actions/runs/4592230751/jobs/8142411680?pr=11902 by running `mvn spotless:apply` and committing/pushing these changes.

**Comment by gakkiyomi** [comment]
> 

@korthout i running 'mvn spotless :apply ' but useless , So I use google-java-format plugin handle those errors and how can i trigger ci job run again ?

**Comment by korthout** [comment]
Sorry for keeping you waiting @gakkiyomi 🙇 I restarted the CI, I'll look at the results tomorrow

**Comment by gakkiyomi** [comment]
> 
@korthout need restart cl job again, i'm not clear whether there are any problems other than code style.

**Review by korthout — APPROVED** [review]
⏭️ The commitlint error is a false-positive. This will be fixed in:
- #12423 

🎉 It does not block this PR. So it's time to merge this!

LGTM 👍 

❤️ Thanks so much for your contribution, and sorry for the long wait @gakkiyomi 

bors merge

**Comment by korthout** [comment]
bors retry

**Pull request merged**
